Aztec : Allow selection of additional doc types via doc picker

a8c-sab-trial opened the pull request:
Fixes #9131. This PR modifies the `UIDocumentPickerViewController` instance presented by Aztec to accommodate selection of multiple document types, using a similar technique applied to the [Media Library](https://github.com/wordpress-mobile/WordPress-iOS/pull/9161). Additionally, it adds support for inserting selected documents into posts via Aztec.

To test (manually): On an iOS 11 device (or Simulator), complete the following sequence against a variety of site types (e.g., self-hosted, free WP-hosted, and paid WP-hosted):

- Select a Site.
- Navigate to `Publish/Site Pages : Start A Page : + (Add Content) : ... (Present the Document Picker)`
- Observe that the Browse tab presented in the view controller includes a Select bar button item, and that 
- Validate that multiple selections are first uploaded to the _Media Library_ and then incorporated into the pending post.

Alternatively, review the attached screen shots.
<img width="487" alt="9131_browse" src="https://user-images.githubusercontent.com/38480191/39219986-f2a288dc-47e2-11e8-8e6d-9df0282e1cd9.png">
<img width="487" alt="9131_selected" src="https://user-images.githubusercontent.com/38480191/39219990-f57e9514-47e2-11e8-8aab-2f61afab6698.png">

SergioEstevao commented:
While this work correctly if you stay inside the editor while the upload is happening, this will fail if the user presses Publish while that upload is ongoing (Async mode).

In order to fix that I suggest the following:
 - Check the code on PostCoordinator for the `save` method
 - On that class you will need to change the processor code where a media object is finished upload and it's a document type.
 - At the moment instead of updating the link to the image, you will need to change the `<img>` tag to a '<a>` link tag.
 - Another nice thing to add, is to insert an attachment placeholder (using a document icon) in order to see the progress of the upload of that document. This will also allow retries if the upload document fails.
 - If you add an attachment placeholder you will also sort a potencial bug where people edit content around the link location and change the ranges where they are planning to insert the document.

If you need any more guidance Steve, just give me a shout on Slack.

frosty commented on WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift:
```diff
@@ -2781,19 +2781,48 @@ extension AztecPostViewController {
                 attachment = insertVideoAttachmentWithPlaceholder()
             default:
-                return
+                guard let urlAsset = exportableAsset as? NSURL else { break }
```
From @SergioEstevao's comment it sounds like some of this may need changing anyway, but whatever we keep here I'd probably recommend we extract it to a method, so this case becomes a single line like the `.image` and `.video` cases.

a8c-sab-trial commented:
@SergioEstevao & @frosty : thank you for pointing out some shortcomings with the original PR. It was a bit myopic, only considering the `EditMode.richText`. I have incorporated your feedback via additional commits that do the following:

- Extract method for a placeholder `ImageAttachment` displayed during upload. This works for `.richText` editing mode, and accounts for changing `NSRange`.
- Created a `DocumentUploadProcessor` to account for `EditMode.html` editing mode, and integrated it into both `AztecPostViewController` & `PostCoordinator`.

![9131_visualmode_placeholder](https://user-images.githubusercontent.com/38480191/39336571-180164be-496d-11e8-86c3-02826163966e.png)
![9131_visualmode_uploaded](https://user-images.githubusercontent.com/38480191/39336572-1af2863a-496d-11e8-8184-857499501965.png)
<img width="487" alt="9131_htmlmode_uploaded" src="https://user-images.githubusercontent.com/38480191/39336575-1db0eec0-496d-11e8-9c8a-2c8973150796.png">

I look forward to hearing your feedback.

SergioEstevao commented on WordPress/Classes/ViewRelated/Aztec/Processors/DocumentUploadProcessor.swift:
```diff
@@ -0,0 +1,28 @@
+    private let title: String
+
+    private lazy var processor = HTMLProcessor(tag: "img", replacer: { [remoteURLString, title] img in
+        var html = "<a href=\"\(remoteURLString)\">\(title)</a>"
```
You are missing a check here. If you see the code on `ImgUploadProcessor` you will see that we do this:

```
guard let imgUploadID = img.attributes.named[MediaAttachment.uploadKey], self.mediaUploadID == imgUploadID else {
            return nil
        }
```

This make sure we changing the correct img link, this way you are going to replace any img in the post with the link info.

SergioEstevao commented:
@frosty this is looking good for me, do you want go give it a final look?

a8c-sab-trial commented:
Thanks for the feedback @frosty; I appreciate your attention to detail. 

I have updated the PR with the following changes:

- I have transitioned to use the `.pages` Gridicon type.
- I have modified `AztecPostViewController.placeholderImage(for:)` to ensure the correct placeholder image is displayed. A better approach would probably involve subclassing either `MediaAttachment` or `ImageAttachment`, but that would require modification of the Pod.
- I'll stand by to hear from either you or @SergioEstevao re: potentially handling intelligent wrapping multiple link insertion.

designsimply commented:
<blockquote>I think we can tackle the formatting issue (if we decide it is an issue) in a separate PR.</blockquote>

I found the formatting pretty awkward 🙂 and opened an enhancement request about it in https://github.com/wordpress-mobile/WordPress-iOS/issues/9293.
